### ISSUE TITLE
DNS resolution is currently unavailable

### ISSUE THREAD

**Issue by beachfork** [pi-hole/pi-hole#1800]
**In raising this issue, I confirm the following:** `{please fill the checkboxes, e.g: [X]}`

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] The issue I am reporting can be *replicated*.
- [X] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:**

1

---
**Expected behaviour:**

The `pihole -g` command would run and finish without errors.

**Actual behaviour:**

When running `pihole -g` I get the following message:
```
[✗] DNS resolution is currently unavailable
[i] Waiting up to 118 seconds before continuing
```

**Steps to reproduce:**

Just run `pihole -g`

**Debug token provided by [uploading `pihole -d` log](https://discourse.pi-hole.net/t/the-pihole-command-with-examples/738#debug):**

rvbvhp661q

**Troubleshooting undertaken, and/or other relevant information:**

Tried a clean install with no success. Thanks with any help!!!

**Comment by dschaper** [comment]
Thank you for the report, we are looking in to this.

**Comment by ookangzheng** [comment]
Me too. I've clean install on Ubuntu 
Your debug token is: da7i0s6q5f

**Comment by dschaper** [comment]
Thank you, I think we have narrowed down where this is happening and I have opened an issue to try and resolve it.

**Comment by dschaper** [comment]
https://github.com/pi-hole/pi-hole/issues/1806

**Comment by RexBelli** [comment]
Just wanted to add my own experience.  I got that error doing a `pihole -up`, however subsequent `pihole -up`s say all is good.  No idea what's going on.

```
::: Checking for updates...
::: Pi-hole Core:   update available
::: FTL:            update available
::: Web Interface:  update available
:::
::: Updating Pi-hole core and web admin files
:::
::: Checking for existing repository...
:::    Updating repo in /etc/.pihole... done!
 done!

        .;;,.
        .ccccc:,.
         :cccclll:.      ..,,
          :ccccclll.   ;ooodc
           'ccll:;ll .oooodc
             .;cll.;;looo:.
                 .. ','.
                .',,,,,,'.
              .',,,,,,,,,,.
            .',,,,,,,,,,,,....
          ....''',,,,,,,'.......
        .........  ....  .........
        ..........      ..........
        ..........      ..........
        .........  ....  .........
          ........,,,,,,,'......
            ....',,,,,,,,,,,,.
               .',,,,,,,,,'.
                .',,,,,,'.
                  ..'''.


  [✓] Root user check
  [i] Performing unattended setup, no whiptail dialogs will be displayed
  [✓] Disk space check

  [✓] Update local cache of available packages

  [✓] Checking apt-get for upgraded packages... 3 updates available
  [i] It is recommended to update your OS after installing the Pi-hole!

  [i] Installer Dependency checks...
  [✓] Checking for apt-utils
  [✓] Checking for dialog
  [✓] Checking for debconf
  [✓] Checking for dhcpcd5
  [✓] Checking for git
  [✓] Checking for iproute2
  [✓] Checking for whiptail


  [i] SELinux mode detected: Disabled
  [✓] Check for existing repository in /etc/.pihole
  [✓] Update repo in /etc/.pihole

  [✓] Check for existing repository in /var/www/html/admin
  [✓] Update repo in /var/www/html/admin

  [i] Main Dependency checks...
  [✓] Checking for bc
  [✓] Checking for cron
  [✓] Checking for curl
  [✓] Checking for dnsmasq
  [✓] Checking for dnsutils
  [✓] Checking for iputils-ping
  [✓] Checking for lsof
  [✓] Checking for netcat
  [✓] Checking for sudo
  [✓] Checking for unzip
  [✓] Checking for wget
  [✗] Checking for idn2 (will be installed)
  [✓] Checking for lighttpd
  [✓] Checking for php-common
  [✓] Checking for php-cgi
  [✗] Checking for php-sqlite3 (will be installed)
  [✓] Installing scripts from /etc/.pihole

  [i] Installing configs from /etc/.pihole...
  [i] Existing dnsmasq.conf found... it is not a Pi-hole file, leaving alone!
  [✓] Copying 01-pihole.conf to /etc/dnsmasq.d/01-pihole.conf

  [i] Creating log and changing owner to dnsmasq... log already exists!

  [i] Installing blocking page...
  [✓] Creating directory for blocking page, and copying files
  [✗] Backing up index.lighttpd.html
      No default index.lighttpd.html file found... not backing up

  [✓] Installing sudoer file

  [✓] Installing latest Cron script

  [✓] Installing latest logrotate script

  [i] FTL Checks...
  [✓] Detected x86_64 architecture
  [i] Checking for existing FTL binary...
  [✓] Downloading and Installing FTL
  [i] Restarting services...

  [✓] Starting dnsmasq service

  [✓] Enabling dnsmasq service to start on reboot

  [✓] Starting lighttpd service

  [✓] Enabling lighttpd service to start on reboot

  [✓] Starting pihole-FTL service

  [✓] Enabling pihole-FTL service to start on reboot

  [i] Preparing to run gravity.sh to refresh hosts...
  [i] Cleaning up previous install (preserving whitelist/blacklist)
  [i] Running gravity.sh
  [✗] DNS resolution is currently unavailable
  [✗] DNS resolution is not available
Unable to complete update, contact Pi-hole
rex@r:~$ nslookup www.google.com
Server:         127.0.0.1
Address:        127.0.0.1#53

Non-authoritative answer:
Name:   www.google.com
Address: 172.217.6.68

rex@r:~$ pihole -up
  [i] Checking for updates...
  [i] Pi-hole Core:     up to date
  [i] FTL:              up to date
  [i] Web Interface:    up to date

  [✓] Everything is up to date!
```

**Comment by Happyfeet01** [comment]
Have the same issue. Is there any solution?

```
root@odroid64:~# pihole -g
  [✗] DNS resolution is currently unavailable
  [i] Waiting up to 119 seconds before continuing...^C

  [i] User-abort detected
  [✓] Cleaning up stray matter
  [✓] DNS service is running
  [✓] Pi-hole blocking is Enabled
root@odroid64:~# pihole -up
  [i] Checking for updates...
fatal: unable to access 'https://github.com/pi-hole/pi-hole.git/': Could not resolve host: github.com
  [i] Pi-hole Core:     up to date
  [i] FTL:              update available

  [i] FTL out of date

  [i] FTL Checks...
  [✓] Detected ARM-aarch64 architecture
  [i] Checking for existing FTL binary...
  [✗] Downloading and Installing FTL
  Error: Unable to get latest release location from GitHub
```

**Comment by viperdriver2000** [comment]
@Happyfeet01 
whats in your `/etc/resolf.conf `?
only `nameserver 127.0.0.1` ?
change it to 
```
nameserver 8.8.8.8
nameserver 8.8.4.4
```
then you can resolve github.com.

**Comment by DiehrWolf** [comment]
I had the opposite resolution.
my /etc/resolf.conf already contained:
`nameserver 8.8.8.8
nameserver 8.8.4.4`

Everything started working once I changed it to:
`nameserver 127.0.0.1`

**Comment by dschaper** [comment]
Closed via #1856

**Comment by Kiina** [comment]
Just in case anyone on arch has the same problem and finds this issue:

Edit your /etc/hosts with `127.0.0.1 pi.hole` and it will work again

**Comment by poldim** [comment]
Just echoing that adding `127.0.0.1      pi.hole` resolved this for me.

Since I'm in a docker, I had to:

`apt-get update
apt-get install nano -y
nano etc/hosts`

**Comment by sanbrother** [comment]
This line in /etc/resolv.conf seems to be the magic
`
nameserver 127.0.0.1
`

Well, I am running armbian, the essential cause is that, pi-hole assigns static ip address in a 'old' way. I manually configured static ip and name server by 'nmtui' command. And then reboot, the problem is gone.

**Comment by jwallet** [comment]
in my case i had this error when my nordvpn was running, i changed some config, set the dns in my vpn to localhost and using tunneling for apps that need the vpn

**Comment by rca** [comment]
> This line in /etc/resolv.conf seems to be the magic
> `nameserver 127.0.0.1`

I ran into this as well, and ensuring that `nameserver 127.0.0.1` is the first item on the list fixed my problem!  The message "DNS resolution is currently unavailable" is sort of ambiguous given that Pihole is the DNS resolver.

Just in case anyone else runs into this, I want to explain how I got into this situation to begin with.

**TL;DR**: set `DNS=127.0.0.1` and `DNSStubListener=no` in `/etc/systemd/resolved.conf`.

I am running Pihole in a Docker container on an Ubuntu 20.04 system, which uses systemd and systemd-resolved.  The systemd resolver needs to be disabled to release port 53.  This can be configured in `/etc/systemd/resolved.conf` with the line: 

```
DNSStubListener=no
```

In addition to setting the above, the file `/etc/resolv.conf` needs to be properly linked to the following with the command `ln -sf /run/systemd/resolve/resolv.conf /etc/resolv.conf`:

```
# ls -al /etc/resolv.conf
lrwxrwxrwx 1 root root 32 Mar 14 02:43 /etc/resolv.conf -> /run/systemd/resolve/resolv.conf
```

**However**, setting `DNSStubListener` alone has implications.  The system will then use the nameservers advertised by the DHCP server. _I configured my DHCP to point to Pihole as the primary DNS server, which prevents Pihole from starting up_.  For me things broke only after trying to restart Pihole, which was way after configuring DHCP.  🤦🏽‍♂️

The fix is easy; add the line `DNS=127.0.0.1` in `/etc/systemd/resolved.conf`, which forces local DNS to be checked prior to upstream servers.

Hope this helps someone else that lands here.

**Comment by aperullo** [comment]
> Clutching at straws here.. can we please get some outputs of:
> 
> ```
> cat /var/log/pihole_updateGravity.log
> ```
@PromoFaux 
```
  [✗] DNS resolution is currently unavailable
  [i] Time until retry: 52
```
Well I suppose that explains why I have zero entries for things being blocked by gravity.

> Does the cli tools work? e.g pihole -g/pihole -w /pihole -b, is it just using the web interface that is facing issues?

Attempting `pihole -g` dies after timing out and the logs are
```
  [✗] DNS resolution is currently unavailable
  [✗] DNS resolution is not available
```
`pihole -w` and `pihole -b` 

Also on starting a fresh container using the known broken compose, these lines appeared in my pihole log
```
  Pi-hole version is v5.6 (Latest: ERROR)
  AdminLTE version is v5.8 (Latest: ERROR)
  FTL version is v5.11 (Latest: ERROR)
```
Though this might just be a red herring since I'm debugging the local dns server and that probably requires a lookup.

> Is it something that the host machine is doing? I've seen a couple of people mention synology above

This is all on a vanilla raspberry pi 400 running rapios_arm64-2021-11-08.


>Restarting the containers and updating gravity has no affect on the permissions once I perform the initial fix.

@lightswitch05
Would you verify which image you're using against mine? Its weird you don't see the problem recur on startup when I do, if we're both running the latest version. I don't think we've quite isolated all the factors yet.

```
> docker inspect pihole | grep Image
"Image": "sha256:72ba441b7847aec62b3f6b6cc38d83e1065f9b4074f9b8fafa73129b5ec31c77"
```

> **notice the invalid entry for DNS1**

Interesting. The following are entries I had during some of these test; but I have been able to recreate this behavior without them so I don't believe they are related. For completeness sake I've included them. They are for a reverse proxy container

`/etc/dnsmasq.d/05-pihole-custom-cname.conf`
```
cname=home.raspberrypi.local,raspberrypi.local
cname=dns.raspberrypi.local,raspberrypi.local
```

`/etc/pihole/custom.list`
```
192.168.0.9 raspberrypi.local
```

⭐ ⭐ ⭐ Below is the big insight for tonight ⭐ ⭐ ⭐ 

> Also, doing a volume mount is still required to reproduce it - not sure why.

This appears to **be specific to bind-mounts** (volumes of the form 'host/path:container/path') which is extra freaky. This docker-compose using named-volumes does not have the issue, even after multiple restarts.

```
  pihole:
    container_name: pihole
    image: pihole/pihole:latest
    restart: unless-stopped
    ports:
      - "53:53/tcp"
      - "53:53/udp"
      - "67:67/udp"
      - "90:80/tcp"
    environment:
      TZ: 'America/Chicago'
    volumes:
      - 'etc-pihole:/etc/pihole/'
      - 'etc-dnsmasq:/etc/dnsmasq.d/'

volumes:
  etc-pihole:
  etc-dnsmasq:
```

However, all is not well when using named volumes, at least for me, `cat /var/log/pihole_updateGravity.log` still doesn't work
```
  [✗] DNS resolution is currently unavailable
  [i] Time until retry:
```
And regardless using a bind-mount is preferable to me and probably others because it lets me version control my `custom.list` and `pihole-custom-cname.conf` among other files. So I'd still call named-volumes a work-around rather than a proper fix.

I hope some of these details are helpful.

**Comment by riesenstiefel** [comment]
I'm having the same issue, running PiHole in a docker container on a Pi 4 and just wanted to share the following oneliner as a quick & dirty workaround, in case it helps somebody:

```bash
docker-compose exec pihole sh -c 'echo "nameserver 127.0.0.1" | sudo tee /etc/resolv.conf > /dev/null'
```
